hnn_core/opt_toy_example/general.py

```diff
@@ -0,0 +1,344 @@
+       fixed_params : unchanging network parameters 
+       predicted_params : the parameters predicted by the optimizer
+    
+       Returns
+       -------
+       net : Network object
+    """
+    
+    net_new = net.copy()
+    # remove existing drives, re-set them with updated parameters 
+    for drive_name in net.external_drives.keys():
+        # clear drive 
+        del net_new.external_drives[drive_name] 
+        # clear connectivity
+        conn_idxs = pick_connection(net_new, src_gids = drive_name)
+        net_new.connectivity = [conn for conn_idx, conn in enumerate(net_new.connectivity) if conn_idx not in conn_idxs]
+        net_new.add_evoked_drive(name = drive_name,
+                              mu = predicted_params[drive_name]['mu'], 
+                              sigma = predicted_params[drive_name]['sigma'], 
+                              numspikes = fixed_params[drive_name]['numspikes'],
+                              location = fixed_params[drive_name]['location'],
+                              n_drive_cells = fixed_params[drive_name]['n_drive_cells'],
+                              cell_specific = fixed_params[drive_name]['cell_specific'],
+                              weights_ampa = predicted_params[drive_name]['ampa_weights'], 
+                              weights_nmda = predicted_params[drive_name]['nmda_weights'], 
+                              space_constant = fixed_params[drive_name]['space_constant'],
+                              synaptic_delays = fixed_params[drive_name]['synaptic_delays'],
+                              probability = fixed_params[drive_name]['probability'],
+                              event_seed = fixed_params[drive_name]['event_seed'],
+                              conn_seed = fixed_params[drive_name]['conn_seed'])
+    return net_new
+
+def _get_opt_net(opt_params, fixed_params):
+    """Returns the optimized network net_."""
+    return 
+
+# These 2 functions will go in a file called metrics.py
+def _rmse_evoked(net, target_statistic, window_len, **params): # params_cobyla = None
+    """The objective function for evoked responses.
+        
+       Parameters
+       -----------
+       net : the Network object 
+       target_dpl : the recorded dipole
+       params : the constraints 
+       
+       Returns
+       -------
+       rmse : normalized root mean squared error between recorded and simulated dipole 
+    """
+
+    fixed_params = _get_fixed_params(net)
+    # get predicted params
+        # gp_minimize & fmin_cobyla return predicted parameters that are formatted differently (bc we are using gp_minimize w/ named args),
+        # we will probably need 2 different _get_predicted_params functions
+    predicted_params = _get_predicted_params(net, **params) 
+    # get network with predicted params 
+    new_net = _set_params(net, fixed_params, predicted_params)
+    # simulate dipole
+    dpl = simulate_dipole(new_net, tstop = 100, n_trials = 1)[0]
```

Comment:
Why do `_get_fixed_params()` and `_get_predicted_params()` need to be separate functions? This should be setup with as much flexibility as possible so that you aren't hardcoding which parameter types are deemed fixed. All fixed parameters should already be instantiated in `net`, so all you have to do is assemble the varied parameters selected by the optimizer, set them within `net` (for an exogenous drive that is targeted for optimization, this will involve overwriting it with the updated parameters), and then running the simulation.